core/drivers/imx_snvs.c

```diff
@@ -2,37 +2,191 @@
 /*
  * Copyright (C) 2020 Pengutronix
  * Rouven Czerwinski <entwicklung@pengutronix.de>
+ * Copyright 2022 NXP
  */
 
-#include <io.h>
 #include <drivers/imx_snvs.h>
+#include <io.h>
 #include <mm/core_memprot.h>
 #include <mm/core_mmu.h>
 #include <stdint.h>
+#include <tee/tee_fs.h>
 #include <types_ext.h>
 #include <trace.h>
 
-enum snvs_security_cfg snvs_get_security_cfg(void)
+#define SNVS_HPLR   0x00
+#define SNVS_HPCOMR 0x04
+#define SNVS_HPSR   0x14
+#define SNVS_LPLR   0x34
+#define SNVS_LPMKCR 0x3C
+
+#define HPSR_SSM_ST_MASK  GENMASK_32(11, 8)
+#define HPSR_SSM_ST_SHIFT 8
+
+#define SNVS_HPSR_SYS_SECURITY_CFG_OFFSET 12
+
+#define SNVS_HPSR_OTPMK_SYND	      GENMASK_32(24, 16)
+#define SNVS_HPSR_OTPMK_ZERO	      BIT(27)
+
+#define SNVS_HPLR_MKS_SL BIT32(9)
+
+#define SNVS_LPLR_MKS_HL BIT32(9)
+
+#define SNVS_HPCOMR_MKS_EN   BIT32(13)
+#define SNVS_HPCOMR_NPSWA_EN BIT32(31)
+
+#define SNVS_LPMKCR_MKCR_MKS_SEL GENMASK_32(1, 0)
+
+enum snvs_ssm_mode {
+	SNVS_SSM_MODE_INIT,
+	SNVS_SSM_MODE_HARD_FAIL,
+	SNVS_SSM_MODE_SOFT_FAIL = 3,
+	SNVS_SSM_MODE_INIT_INTERMEDIATE = 8,
+	SNVS_SSM_MODE_CHECK,
+	SNVS_SSM_MODE_NON_SECURE = 11,
+	SNVS_SSM_MODE_TRUSTED = 13,
+	SNVS_SSM_MODE_SECURE = 15,
+};
+
+enum snvs_security_cfg {
+	SNVS_SECURITY_CFG_FAB,
+	SNVS_SECURITY_CFG_OPEN,
+	SNVS_SECURITY_CFG_CLOSED,
+	SNVS_SECURITY_CFG_FIELD_RETURN,
+};
+
+/*
+ * Return true if the master key is OTPMK, false otherwise.
+ */
+static bool is_otpmk_selected(void)
+{
+	uint32_t hp_mks = 0;
+	vaddr_t base = core_mmu_get_va(SNVS_BASE, MEM_AREA_IO_SEC, SNVS_SIZE);
+
+	hp_mks = io_read32(base + SNVS_HPCOMR);
+
+	/*
+	 * The master key selection might be done by the MASTER_KEY_SEL field
+	 * of LPMKCR instead.
+	 */
+	if (hp_mks & SNVS_HPCOMR_MKS_EN) {
+		uint32_t lp_mks = io_read32(base + SNVS_LPMKCR);
+
+		if (lp_mks & SNVS_LPMKCR_MKCR_MKS_SEL)
+			return false;
+	}
+
+	return true;
+}
+
+/*
+ * Return true if the master key selection is locked, false otherwise.
+ */
+static bool is_mks_locked(void)
+{
+	uint32_t hp_lr = 0;
+	uint32_t lp_lr = 0;
+	vaddr_t base = core_mmu_get_va(SNVS_BASE, MEM_AREA_IO_SEC, SNVS_SIZE);
+
+	hp_lr = io_read32(base + SNVS_HPLR);
+	lp_lr = io_read32(base + SNVS_LPLR);
+
+	if ((hp_lr & SNVS_HPLR_MKS_SL) || (lp_lr & SNVS_LPLR_MKS_HL))
+		return true;
+
+	return false;
+}
+
+/*
+ * Set the Master key to use OTPMK and lock it. Return true if the operation is
+ * successful. Return false if the master key is already locked.
```

Comment:
This does not return anything, function is `static void`, was the intention to check something here or was this changed during refactoring?

Reply:
It was changed during the review. Thanks for noticing it